Improve launch time of WebKit processes, v2

pvollan opened the pull request:
#### 9590b618193248a44916c41e66a6668cc070e20d
<pre>
Improve launch time of WebKit processes, v2
<a href="https://bugs.webkit.org/show_bug.cgi?id=261966">https://bugs.webkit.org/show_bug.cgi?id=261966</a>
rdar://problem/115906222

Reviewed by Brent Fulgham.

To benefit of the launch time improvements by launching WebKit processes as extensions, this patch implements
this approach by using ExtensionKit for launching. The feature is currently behind a flags, which is not
being turned on in this patch. The extensions are currently being installed in the WebKit framework.

* Source/WebKit/Configurations/BaseExtension.xcconfig:
* Source/WebKit/Platform/spi/Cocoa/ExtensionKitSPI.h: Copied from Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtensionBridge.h.
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.WebContent.sb.in:
* Source/WebKit/Scripts/process-entitlements.sh:
* Source/WebKit/Shared/API/Cocoa/WKMain.h:
* Source/WebKit/Shared/API/Cocoa/WKMain.mm:
(WKExtensionEventHandler):
(WKExtensionMain): Deleted.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtension.entitlements: Removed.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtension.swift:
(GPUProcessExtension.handle(_:)):
(AuxiliaryProcessExtension.main): Deleted.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtensionBridge.h:
* Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtensionBridge.mm:
(handleNewConnection):
(extensionMain): Deleted.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/GPUProcessExtension.swift: Renamed from Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtension.swift.
(GPUProcessExtension.handle(_:)):
* Source/WebKit/Shared/AuxiliaryProcessExtensions/NetworkingProcessExtension.swift: Copied from Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtension.swift.
(NetworkingProcessExtension.handle(_:)):
* Source/WebKit/Shared/AuxiliaryProcessExtensions/WebContentProcessExtension.swift: Copied from Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtension.swift.
(WebContentProcessExtension.handle(_:)):
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceEntryPoint.h:
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
(WebKit::XPCServiceEventHandler):
(WebKit::XPCServiceMain):
* Source/WebKit/UIProcess/Launcher/ProcessLauncher.h:
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::serviceNameAndIdentifier):
(WebKit::launchWithExtensionKit):
(WebKit::ProcessLauncher::launchProcess):
(WebKit::ProcessLauncher::finishLaunchingProcess):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/268990@main">https://commits.webkit.org/268990@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7755c874e0242a9149ee7d8b20865aace99e7fa6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21156 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21509 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22213 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23027 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19662 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24775 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21706 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20898 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21380 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21081 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18343 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23880 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/21342 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18238 "Passed tests") | | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19319 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19375 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19908 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16926 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19190 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5090 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23476 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19772 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtensionBridge.h:
```diff
@@ -25,11 +25,14 @@
 #pragma once
 
+#import <objc/objc.h>
+#import <xpc/xpc.h>
```
I wonder if we could do these imports in the implementation files, since the new declaration just needs a definition for `xpc_conneciton_t`, which we could forward declare. I think these are fairly big headers, so avoiding them in this header could be a nice build-time benefit.

pvollan replied:
That is a good point! I tried doing the forward declaration, although it became a bit complicated since the definition of `xpc_connection_t` depends on the language. If you are ok with it, I'll keep the include for now.

brentfulgham commented on Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtensionBridge.mm:
```diff
@@ -27,7 +27,7 @@
 
 #import <WebKit/WKMain.h>
```
Maybe add the `objc` include headers here instead of in the `AuxiliaryProcessExtensionBridge.h` header.

pvollan replied:
I removed the objc header in the latest patch, since it was not required.

brentfulgham commented on Source/WebKit/Shared/AuxiliaryProcessExtensions/GPUProcessExtension.swift:
```diff
@@ -23,11 +23,17 @@
-    static func main() throws -> Void {
-        extensionMain(CommandLine.argc, CommandLine.unsafeArgv)
+class GPUProcessExtension {
+
```
Nit: Unneeded whitespace.

brentfulgham commented on Source/WebKit/Shared/AuxiliaryProcessExtensions/NetworkingProcessExtension.swift:
```diff
@@ -0,0 +1,39 @@
+
+@main
+class NetworkingProcessExtension {
+
```
Ditto extra newline.

brentfulgham commented on Source/WebKit/Shared/AuxiliaryProcessExtensions/WebContentProcessExtension.swift:
```diff
@@ -0,0 +1,39 @@
+
+@main
+class WebContentProcessExtension {
+
```
Ditto.

pvollan replied:
Fixed in all 3 places in latest patch.

brentfulgham commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -56,8 +56,66 @@
 namespace WebKit {
 
+#if USE(EXTENSIONKIT)
+static std::pair<const char*, RetainPtr<NSString>> serviceNameAndIdentifier(ProcessLauncher::ProcessType processType)
```
Should this maybe just be an Objective C NSDictionary, since it's only really needed for Cocoa APIs?

brentfulgham replied:
Actually, maybe not since the key is used for regular WebKit C++ code.

cdumez replied:
Please use ASCIILiteral instead of `const char*`, then thread-safety is not a concern.

pvollan replied:
That is a good point! I will change this to `ASCIILiteral`.

Thanks for reviewing!

brentfulgham commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -82,17 +140,40 @@
+            return;
+        if (error)
+            NSLog(@"Error launching process %@ error %@", process, error);
+        callOnMainRunLoop([weakProcessLauncher = weakProcessLauncher, name = name, process = RetainPtr<_SEExtensionProcess>(process)] {
```
The use of `callOnMainRunLoop` makes it seem like some of this code might run off the main thread. Do we need to be doing an isolated copy of name here? Or adding asserts that the other code referencing this `name` (which comes from `serviceNameAndIdentifier`) is on the right thread?

brentfulgham commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -56,8 +56,66 @@
+        auto block = [handler = handler, weakProcessLauncher = ThreadSafeWeakPtr { processLauncher }, name = name](_SEContentProcess* _Nullable process, NSError* _Nullable error) {
+            handler(weakProcessLauncher, process, name, error);
+        };
+        [manager performSelector:@selector(contentProcessWithConfiguration:completion:) withObject:configuration.get() withObject:block];
```
If `manager` does this work off the main thread, we might need to use isolated copy on `name`.

brentfulgham replied:
Ditto for the other `case` sections.

brentfulgham reviewed (approved):
Looks good. Can you double-check the `const char*` handling of `name` for the cross-thread issues I asked about? r=me

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -82,17 +140,40 @@
+            NSLog(@"Error launching process %@ error %@", process, error);
+        callOnMainRunLoop([weakProcessLauncher = weakProcessLauncher, name = name, process = RetainPtr<_SEExtensionProcess>(process)] {
+            auto connection = [process makeLibXPCConnectionError:nil];
+            weakProcessLauncher.get()->m_xpcConnection = connection;
```
You should use a local variable for the result of `weakProcessLauncher.get()`. Calling `weakProcessLauncher.get()` is as cheap as you might expect in this case. So calling it 3 times in a row is not ideal.

pvollan replied:
That is a good point! Will fix in upcoming patch.

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -82,17 +140,40 @@
+        callOnMainRunLoop([weakProcessLauncher = weakProcessLauncher, name = name, process = RetainPtr<_SEExtensionProcess>(process)] {
+            auto connection = [process makeLibXPCConnectionError:nil];
+            weakProcessLauncher.get()->m_xpcConnection = connection;
+            weakProcessLauncher.get()->m_process = process;
```
Can we `WTFMove(process)` ?

pvollan replied:
Yes, will fix in upcoming patch.

pvollan committed:
Improve launch time of WebKit processes, v2
https://bugs.webkit.org/show_bug.cgi?id=261966
rdar://problem/115906222

Reviewed by Brent Fulgham.

To benefit of the launch time improvements by launching WebKit processes as extensions, this patch implements
this approach by using ExtensionKit for launching. The feature is currently behind a flags, which is not
being turned on in this patch. The extensions are currently being installed in the WebKit framework.

* Source/WebKit/Configurations/BaseExtension.xcconfig:
* Source/WebKit/Platform/spi/Cocoa/ExtensionKitSPI.h: Copied from Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtensionBridge.h.
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.WebContent.sb.in:
* Source/WebKit/Scripts/process-entitlements.sh:
* Source/WebKit/Shared/API/Cocoa/WKMain.h:
* Source/WebKit/Shared/API/Cocoa/WKMain.mm:
(WKExtensionEventHandler):
(WKExtensionMain): Deleted.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtension.entitlements: Removed.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtension.swift:
(GPUProcessExtension.handle(_:)):
(AuxiliaryProcessExtension.main): Deleted.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtensionBridge.h:
* Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtensionBridge.mm:
(handleNewConnection):
(extensionMain): Deleted.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/GPUProcessExtension.swift: Renamed from Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtension.swift.
(GPUProcessExtension.handle(_:)):
* Source/WebKit/Shared/AuxiliaryProcessExtensions/NetworkingProcessExtension.swift: Copied from Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtension.swift.
(NetworkingProcessExtension.handle(_:)):
* Source/WebKit/Shared/AuxiliaryProcessExtensions/WebContentProcessExtension.swift: Copied from Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtension.swift.
(WebContentProcessExtension.handle(_:)):
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceEntryPoint.h:
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
(WebKit::XPCServiceEventHandler):
(WebKit::XPCServiceMain):
* Source/WebKit/UIProcess/Launcher/ProcessLauncher.h:
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::serviceNameAndIdentifier):
(WebKit::launchWithExtensionKit):
(WebKit::ProcessLauncher::launchProcess):
(WebKit::ProcessLauncher::finishLaunchingProcess):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/268990@main
